201810/raw-tcp-traffic-shaping-with-istio-1.1.0.md

```diff
@@ -0,0 +1,403 @@
+        image: vnoronha/tcp-echo-server:latest
+        args: [ "9000", "two" ] # prefix: two
+        imagePullPolicy: IfNotPresent
+        ports:
+        - containerPort: 9000
+```
+
+### 部署Minikube
+
+[Minikube](https://kubernetes.io/docs/setup/minikube/)是Kubernetes本地开发的最佳工具。用下面的命令启动Minikube实例：
+
+```bash
+$ minikube start --bootstrapper kubeadm       \
+                 --memory=8192                \
+                 --cpus=4                     \
+                 --kubernetes-version=v1.10.0 \
+                 --vm-driver=virtualbox
+Starting local Kubernetes v1.10.0 cluster...
+...
+Kubectl is now configured to use the cluster.
+Loading cached images from config file.
+```
+
+### 安装 Istio
+
+在撰写本文时，Istio 1.1.0还没有发布。因此我使用了Istio的 [Daily Pre-Release](https://gcsweb.istio.io/gcs/istio-prerelease/daily-build/master-20181017-09-15/)来演示这个新特性。请参考[Istio文档](https://istio.io/docs/setup/kubernetes/download-release/)学习下载和配置Istio。
+
+一旦部署完毕，有一个简单的方法来完全部署Istio：
+
+```bash
+$ kubectl apply -f install/kubernetes/helm/istio/templates/crds.yaml
+customresourcedefinition.apiextensions.k8s.io/virtualservices.networking.istio.io created
+...
+customresourcedefinition.apiextensions.k8s.io/templates.config.istio.io created
+customresourcedefinition.apiextensions.k8s.io/handlers.config.istio.io created
+
+$ kubectl apply -f install/kubernetes/istio-demo.yaml
+namespace/istio-system created
+...
+destinationrule.networking.istio.io/istio-policy created
+destinationrule.networking.istio.io/istio-telemetry created 
+```
+
+### 使用Istio代理部署TCP Echo服务
+
+为了演示Istio的路由机制，我们以side-car模式部署`tcp-echo-server` ：
+
+```bash
+$ kubectl apply -f <(istioctl kube-inject -f service.yaml)
+service/tcp-echo-server created
+deployment.extensions/tcp-echo-server-v1 created
+deployment.extensions/tcp-echo-server-v2 created
+```
+
+通过下面的命令来验证服务运行：
+
+```bash
+$ kubectl get pods
+NAME                                  READY     STATUS    RESTARTS   AGE
+tcp-echo-server-v1-78684f5697-sv5r5   2/2       Running   0          56s
+tcp-echo-server-v2-74bf9999c8-hhhf9   2/2       Running   0          56s
+
+$ kubectl logs tcp-echo-server-v1-78684f5697-sv5r5 tcp-echo-server
+listening on [::]:9000, prefix: one
+
+$ kubectl logs tcp-echo-server-v2-74bf9999c8-hhhf9 tcp-echo-server
+listening on [::]:9000, prefix: two
+```
+
+## Istio加权TCP路由
+
+这是本练习的最后一部分，定义`VirtualService`，`DestinationRule`和带有权重路由的 `Gateway`，并验证系统行为。
+
+### 路由配置
+
+创建带有两个`subset`的`DestinationRule`来代表两个版本的 TCP Echo服务。`Gateway` 容许流量通过端口`9000`访问服务。最后，`VirtualService`限定了80%的流量必须被路由到TCP Echo服务的v1版本，20%被路由到v2版本。
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: destination
+spec:
+  host: tcp-echo-server
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: gateway
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - port:
+      number: 9000
```

Comment:
Also, this has been updated: https://github.com/venilnoronha/venilnoronha.github.io/blob/source/_posts/2018-10-19-raw-tcp-traffic-shaping-with-istio-1.1.0.md#routing-configuration